server/sql_info_fetcher.go

```diff
@@ -0,0 +1,325 @@
+				fmt.Fprintf(fw, "%s\n", strings.Join(row, "\t"))
+			}
+		case <-timer.C:
+			cancelFunc()
+		}
+		err = <-errChan
+		if err != nil {
+			err = sh.writeErrFile(zw, "profile.err.txt", err)
+			terror.Log(err)
+			return
+		}
+	}
+}
+
+func (sh *sqlInfoFetcher) writeErrFile(zw *zip.Writer, name string, err error) error {
+	fw, err1 := zw.Create(name)
+	if err1 != nil {
+		return err1
+	}
+	fmt.Fprintf(fw, "error: %v", err)
+	return nil
+}
+
+type explainAnalyzeResult struct {
+	rows [][]string
+	err  error
+}
+
+func (sh *sqlInfoFetcher) getExplainAnalyze(ctx context.Context, sql string, resultChan chan<- *explainAnalyzeResult) {
+	recordSets, err := sh.s.(sqlexec.SQLExecutor).Execute(ctx, fmt.Sprintf("explain analyze %s", sql))
+	if len(recordSets) > 0 {
+		defer terror.Call(recordSets[0].Close)
+	}
+	if err != nil {
+		resultChan <- &explainAnalyzeResult{err: err}
+		return
+	}
+	rows, err := testkit.ResultSetToStringSlice(ctx, sh.s, recordSets[0])
+	if err != nil {
+		terror.Log(err)
+		rows = nil
+		return
+	}
+	resultChan <- &explainAnalyzeResult{rows: rows}
+}
+
+func (sh *sqlInfoFetcher) catchCPUProfile(ctx context.Context, sec int, zw *zip.Writer, errChan chan<- error) {
+	// dump profile
+	fw, err := zw.Create("profile")
+	if err != nil {
+		errChan <- err
+		return
+	}
+	if err := pprof.StartCPUProfile(fw); err != nil {
+		errChan <- err
+		return
+	}
+	sleepWithCtx(ctx, time.Duration(sec)*time.Second)
+	pprof.StopCPUProfile()
+	errChan <- nil
+}
+
+func (sh *sqlInfoFetcher) getStatsForTable(pair tableNamePair) (*handle.JSONTable, error) {
+	is := sh.do.InfoSchema()
+	h := sh.do.StatsHandle()
+	tbl, err := is.TableByName(model.NewCIStr(pair.DBName), model.NewCIStr(pair.TableName))
+	if err != nil {
+		return nil, err
+	}
+	js, err := h.DumpStatsToJSON(pair.DBName, tbl.Meta(), nil)
+	return js, err
+}
+
+func (sh *sqlInfoFetcher) getShowCreateTable(pair tableNamePair, zw *zip.Writer) error {
+	recordSets, err := sh.s.(sqlexec.SQLExecutor).Execute(context.TODO(), fmt.Sprintf("show create table `%v`.`%v`", pair.DBName, pair.TableName))
+	if len(recordSets) > 0 {
+		defer terror.Call(recordSets[0].Close)
+	}
+	if err != nil {
+		return err
+	}
+	sRows, err := testkit.ResultSetToStringSlice(context.Background(), sh.s, recordSets[0])
+	if err != nil {
+		terror.Log(err)
+		return nil
+	}
+	fw, err := zw.Create(fmt.Sprintf("%v.%v.schema.txt", pair.DBName, pair.TableName))
+	if err != nil {
+		terror.Log(err)
+		return nil
+	}
+	for _, row := range sRows {
+		fmt.Fprintf(fw, "%s\n", strings.Join(row, "\t"))
+	}
+	return nil
+}
+
+func (sh *sqlInfoFetcher) extractTableNames(sql, curDB string) (map[tableNamePair]struct{}, error) {
+	p := parser.New()
+	charset, collation := sh.s.GetSessionVars().GetCharsetInfo()
+	stmts, _, err := p.Parse(sql, charset, collation)
+	if err != nil {
+		return nil, err
+	}
+	extractor := &tableNameExtractor{
+		curDB: curDB,
+		names: make(map[tableNamePair]struct{}),
+	}
+	for _, stmt := range stmts {
```

Comment:
So we allow multiple statements? Will it cause problems if we use `explain sql`?